File: LRSMessageToolBar/Classes/other/LRSMessageBar.m

#import "LRSMessageBar.h"
#import "LRSMessageToolBarHelper.h"
#import "LRSMessageInputBar.h"
#import "LRSMemePackagesView.h"
//#import "LRSMessageBar-swift.h"

@interface LRSMessageBar () <UITextViewDelegate> {
    BOOL micphonePerPermission_;
}
@property(nonatomic, strong) LRSMessageToolBarConfigure *configure;
@property(nonatomic, strong) LRSMessageInputBar *toolBar;
@property(nonatomic, strong, nullable) LRSMemePackagesView *memePackagesView;
@end

@implementation LRSMessageBar

- (instancetype)initWithConfigure:(LRSMessageToolBarConfigure *)configure {
    if (self = [super init]) {
        self.configure = configure;
        [self layoutSubviews];
    }
    return self;
}

- (void)layoutSubviews {
    [self addSubview:self.toolBar];
}

- (void)sendOut_ {
    NSString *strUrl = [self.toolBar.inputTextView.text stringByReplacingOccurrencesOfString:@" " withString:@""];
    strUrl = [strUrl stringByReplacingOccurrencesOfString:@"\n" withString:@""];
    strUrl = [strUrl stringByReplacingOccurrencesOfString:@"\r" withString:@""];
    if (strUrl.length > 0) {
        if (self.delegate && [self.delegate respondsToSelector:@selector(messageToolBarDidClickedReturn:)]) {
            [self.delegate messageToolBarDidClickedReturn:self];
            self.toolBar.inputTextView.text = @"";
            [self.toolBar.inputTextView setContentOffset:CGPointMake(0, 0) animated:YES];
        }
    } else {
        [self.delegate messageToolBarDidClickedReturn:self];
    }
}

- (CGFloat)textViewContentHeight_:(UITextView *)textView {
    CGFloat height = ceilf((float) [textView sizeThatFits:textView.frame.size].height);
    CGFloat min = self.configure.textViewConfigure.minHeight;
    CGFloat max = self.configure.textViewConfigure.maxHeight;
    return MIN(max, MAX(min, height));
}

#pragma mark - UITextViewDelegate

- (BOOL)textViewShouldBeginEditing:(UITextView *)textView {
    if (self.delegate && [self.delegate respondsToSelector:@selector(messageToolBarInputTextViewShouldBeginEditing)]) {
        return [self.delegate messageToolBarInputTextViewShouldBeginEditing:self];
    }
    return YES;
}

- (void)textViewDidBeginEditing:(UITextView *)textView {
    if (self.delegate && [self.delegate respondsToSelector:@selector(messageToolBarInputTextViewDidBeginEditing:)]) {
        [self.delegate messageToolBarInputTextViewDidBeginEditing:self];
    }
}

- (void)textViewDidEndEditing:(UITextView *)textView {
    if (self.delegate && [self.delegate respondsToSelector:@selector(messageToolBarInputTextViewDidEndEditing:)]) {
        [self.delegate messageToolBarInputTextViewDidEndEditing:self];
    }
    [textView resignFirstResponder];
}

- (void)textViewDidChangeSelection:(UITextView *)textView {
    CGRect r = [textView caretRectForPosition:textView.selectedTextRange.end];
    CGFloat caretY = MAX(r.origin.y - textView.frame.size.height + r.size.height + 8, 0);
    if (textView.contentOffset.y < caretY && r.origin.y != INFINITY) {
        textView.contentOffset = CGPointMake(0, caretY);
    }
}

- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
    if ([text isEqualToString:@"\n"]) {
        [self sendOut_];
        return NO;
    }
    return YES;
}

- (void)textViewDidChange:(UITextView *)textView {
    NSInteger length = self.configure.textViewConfigure.acceptLength;
//    if ([DXMessageToolBarTextHandler textCountWithText:textView.text] > length) {
//        textView.text = [DXMessageToolBarTextHandler handleTextWithText:textView.text maxLength:length];
//    }
//    [self willShowInputTextViewToHeight:[self textViewContentHeight_:self.toolBar.inputTextView]];
}

#pragma mark - UIKeyboardNotification

- (void)keyboardWillChangeFrame:(NSNotification *)notification {
    if (self.hidden) {
        return;
    }
    NSDictionary *userInfo = notification.userInfo;
    CGRect endFrame = [userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
    CGPoint centerEndPoint = [userInfo[UIKeyboardCenterEndUserInfoKey] CGPointValue];
    CGRect beginFrame = CGRectMake(0, self.frame.origin.y + self.frame.size.height, endFrame.size.width, endFrame.size.height);
    CGFloat duration = [userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    UIViewAnimationCurve curve = [userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];

    [UIView animateWithDuration:duration
                          delay:0.0f
                        options:(curve << 16 | UIViewAnimationOptionBeginFromCurrentState)
                     animations:^{
//                         [self willShowKeyboardFromFrame:beginFrame toFrame:endFrame];
                         if (centerEndPoint.y > [LRSMessageToolBarHelper screenHeight] + 20) {
//                             [self willShowBottomView:nil];
                         }
                     } completion:nil];
}

- (void)didClickFaceBtn:(UIButton *)button {
    NSLog(@"点击表情");

}

// 点击转换按钮,控制模式转换
- (void)msgSendClick:(UIButton *)btn {

}

- (void)onImagePickButtonClick:(UIButton *)button {
    if (self.delegate && [self.delegate respondsToSelector:@selector(messageToolBar:buttonDidClicked:)]) {
        [self.delegate messageToolBar:self buttonDidClicked:1];
    }
}
//
//#pragma mark - change frame
//
//- (void)willShowBottomHeight:(CGFloat)tobe_bottomHeight {
//    if ((self.toolBarSkinType == RecreationRoomChatToolBarType || self.toolBarSkinType == RecreationRoomAccompyChatToolBarType)) {
//        CGRect fromFrame = self.frame;
//        CGFloat bottomHeight = (tobe_bottomHeight == 0) ? kSafeAreaBottomHeight : tobe_bottomHeight;
//        CGFloat toHeight = self.toolbarView.frame.size.height + bottomHeight;
//        CGRect toFrame = CGRectMake(fromFrame.origin.x, SCREEN_HEIGHT - toHeight, fromFrame.size.width, toHeight);
//        if (tobe_bottomHeight == 0) {
//            bottomHeight = 0;
//            toFrame = CGRectMake(0, SCREEN_HEIGHT, SCREEN_WIDTH, toHeight);
//        }
//
//        //如果需要将所有扩展页面都隐藏，而此时已经隐藏了所有扩展页面，则不进行任何操作
//        if (tobe_bottomHeight == 0 && self.frame.size.height == self.toolbarView.frame.size.height) {
//            return;
//        }
//
//        if (self.isHidden) {
//        } else {
//            self.frame = toFrame;
//            if (_delegate && [_delegate respondsToSelector:@selector(didChangeFrameToHeight:)]) {
//                [_delegate didChangeFrameToHeight:toHeight];
//            }
//        }
//    } else {
//        CGRect fromFrame = self.frame;
//        CGFloat bottomHeight = (tobe_bottomHeight == 0) ? 0 + kSafeAreaBottomHeight : tobe_bottomHeight;
//        CGFloat toHeight = self.toolbarView.frame.size.height + bottomHeight;
//        CGRect toFrame = CGRectMake(fromFrame.origin.x, fromFrame.origin.y + (fromFrame.size.height - toHeight), fromFrame.size.width, toHeight);
//
//        //如果需要将所有扩展页面都隐藏，而此时已经隐藏了所有扩展页面，则不进行任何操作
//        if (tobe_bottomHeight == 0 && self.frame.size.height == self.toolbarView.frame.size.height) {
//            return;
//        }
//
//        if (self.isHidden) {
//        } else {
//            self.frame = toFrame;
//            if (_delegate && [_delegate respondsToSelector:@selector(didChangeFrameToHeight:)]) {
//                [_delegate didChangeFrameToHeight:toHeight];
//            }
//        }
//    }
//}
//
//- (void)willShowBottomView:(UIView *)bottomView {
//    if (![self.activityBottomView isEqual:bottomView]) {
//        if (self.activityBottomView) {
//            [self.activityBottomView removeFromSuperview];
//        }
//        CGFloat bottomHeight = bottomView ? bottomView.frame.size.height : 0;
//        [self willShowBottomHeight:bottomHeight];
//        if (bottomView) {
//            CGRect rect = bottomView.frame;
//            rect.origin.y = CGRectGetMaxY(self.toolbarView.frame);
//            bottomView.frame = rect;
//            [self addSubview:bottomView];
//        } else {
//
//        }
//        self.activityBottomView = bottomView;
//    }
//}
//
//- (void)willShowKeyboardFromFrame:(CGRect)beginFrame toFrame:(CGRect)toFrame {
//    if (beginFrame.origin.y == [[UIScreen mainScreen] bounds].size.height) {
//        //一定要把self.activityBottomView置为空
//        [self willShowBottomHeight:toFrame.size.height];
//        if (self.activityBottomView) {
//            [self.activityBottomView removeFromSuperview];
//        }
//        self.activityBottomView = nil;
//    } else if (toFrame.origin.y == [[UIScreen mainScreen] bounds].size.height) {
//        [self willShowBottomHeight:0];
//    } else {
//        [self willShowBottomHeight:toFrame.size.height];
//    }
//}
//
//- (void)willShowInputTextViewToHeight:(CGFloat)toHeight {
//    if (toHeight < self.kInputTextViewMinHeight) {
//        toHeight = self.kInputTextViewMinHeight;
//    }
//    if (toHeight > self.maxTextInputViewHeight) {
//        toHeight = self.maxTextInputViewHeight;
//    }
//
//    if (toHeight == self.previousTextViewContentHeight) {
//        return;
//    } else {
//        CGFloat changeHeight = toHeight - self.previousTextViewContentHeight;
//
//        CGRect rect = self.frame;
//        rect.size.height += changeHeight;
//        rect.origin.y -= changeHeight;
//        self.frame = rect;
//
//        rect = self.toolbarView.frame;
//        rect.size.height += changeHeight;
//        self.toolbarView.frame = rect;
//        self.previousTextViewContentHeight = toHeight;
//
//        rect = self.contentView.frame;
//        rect.size.height += changeHeight;
//        self.contentView.frame = rect;
//
//        rect = self.inputBaseView.frame;
//        rect.size.height += changeHeight;
//        self.inputBaseView.frame = rect;
//
//        if (_delegate && [_delegate respondsToSelector:@selector(didChangeFrameToHeight:)] && !self.isHidden) {
//            [_delegate didChangeFrameToHeight:self.frame.size.height];
//        }
//    }
//}
//
//
//#pragma mark - public
//
///**
// *  停止编辑
// */
//- (BOOL)endEditing:(BOOL)force {
//    if (!force && (self.hidden || !self.inputTextView.isFirstResponder)) {
//        return YES;
//    } else {
//        BOOL result = [super endEditing:force];
//        self.faceButton.selected = NO;
//        [self.inputTextView endEditing:YES];
//        if ([self.inputTextView isFirstResponder] && [self.inputTextView canResignFirstResponder]) {
//            [self.inputTextView resignFirstResponder];
//        } else {
//            if (self.inputView.isFirstResponder) {
//                [self.inputTextView resignFirstResponder];
//            }
//        }
//        [self willShowBottomView:nil];
//        return result;
//    }
//}

- (LRSMemePackagesView *)memePackagesView {
    if (!_memePackagesView) {
        _memePackagesView = [[LRSMemePackagesView alloc] initWithFrame:CGRectZero configures:[LRSMessageToolBarHelper allEmojis]];
        __weak typeof(self) weakSelf = self;
        _memePackagesView.itemHandler = ^(LRSMemeSinglePage *view, LRSMemePackageConfigureItem *item) {
            NSString *chatText = weakSelf.toolBar.inputTextView.text;
            weakSelf.toolBar.inputTextView.text = [NSString stringWithFormat:@"%@%@", chatText ?: @"", item.emojiValue];
            [weakSelf textViewDidChange:weakSelf.toolBar.inputTextView];
        };
        _memePackagesView.backspaceHandler = ^(LRSMemeSinglePage *view) {
            [weakSelf.toolBar.inputTextView deleteBackward];
        };
        _memePackagesView.sendOutHandler = ^(LRSMemePackgaesView *view) {
            [weakSelf sendOut_];
        };
    }
    return _memePackagesView;
}

- (LRSMessageInputBar *)toolBar {
    if (!_toolBar) {
        _toolBar = [LRSMessageInputBar toolBarWithConfigure:self.configure];
        _toolBar.backgroundColor = [UIColor whiteColor];
        _toolBar.layer.shadowColor = [[UIColor blackColor] CGColor];
        _toolBar.layer.shadowOffset = CGSizeMake(0, -2 * [LRSMessageToolBarHelper scale]);
        _toolBar.layer.shadowRadius = 5.0 * [LRSMessageToolBarHelper scale];
        _toolBar.layer.shadowOpacity = 0.1;
        _toolBar.inputTextView.delegate = self;
        __weak typeof(self) weakSelf = self;
        _toolBar.recordingBtn.touchBegan = ^(LRSMessageToolBarRecordButton *button) {
            NSLog(@"按下按钮,开始录音");
            micphonePerPermission_ = [self.datasource audioPermission];
            if (!micphonePerPermission_) {
                weakSelf.toolBar.recordingBtn.selected = false;
                return;
            }
            if (weakSelf.delegate && [self.delegate respondsToSelector:@selector(messageToolBarBeganToSpeak:)]) {
                [weakSelf.delegate messageToolBarBeganToSpeak:weakSelf];
            }
        };
        _toolBar.recordingBtn.touchEnd = ^(LRSMessageToolBarRecordButton *button) {
            NSLog(@"结束录音");
            if ([weakSelf.delegate respondsToSelector:@selector(messageToolBarEndSpeaking:)]) {
                [weakSelf.delegate messageToolBarEndSpeaking:weakSelf];
            }
        };
        _toolBar.recordingBtn.dragEnter = ^(LRSMessageToolBarRecordButton *button) {
            NSLog(@"区域划入");
            if ([weakSelf.delegate respondsToSelector:@selector(messageToolBarDragEnterRecordScope)]) {
                [weakSelf.delegate messageToolBarDragEnterRecordScope:weakSelf];
            }
        };
        _toolBar.recordingBtn.dragOutside = ^(LRSMessageToolBarRecordButton *button) {
            NSLog(@"区域划出");
            if ([weakSelf.delegate respondsToSelector:@selector(messageToolBarDragOutRecordScope)]) {
                [weakSelf.delegate messageToolBarDragOutRecordScope:weakSelf];
            }
        };
        _toolBar.recordingBtn.dragOutsideRelease = ^(LRSMessageToolBarRecordButton *button) {
            NSLog(@"区域划出松手");
            if ([weakSelf.delegate respondsToSelector:@selector(messageToolBarSlideTopToCancelRecording:)]) {
                [weakSelf.delegate messageToolBarSlideTopToCancelRecording:weakSelf];
            }
        };
    }
    return _toolBar;
}

- (void)dealloc {

}

@end

File: LRSMessageToolBar/Classes/Core/LRSMessageBar.swift
//
//  LRSMessageBar.swift
//  LRSMessageToolBar
//
//  Created by sama 刘 on 2021/11/9.
//

import UIKit

@objc public class LRSMessageBar: UIView {

    enum Mode {
        case normal
        case keyboard
        case meme
    }

    struct BarConfigure {
        let toolBottomOffset: CGFloat = 0
        let memeAnimationDuration: TimeInterval = 0.3
        let memeAnimationOffset: CGFloat = 50
    }

    @objc
    open weak var delegate: LRSMesssageBarProtocol?

    @objc
    public let toolBar: LRSMessageInputBar

    private let configure: LRSMessageToolBarConfigure

    private lazy var itemHandler: LRSMemePackagesContentView.ItemHandler = {[weak self] page, item in
        self?.append(text: item?.emojiValue)
    }

    private lazy var deleteHandler: LRSMemePackagesView.ItemHandler = {[weak self] _, _ in
        self?.toolBar.inputTextView.deleteBackward()
    }

    private lazy var confirmHandler: LRSMemePackagesView.ItemHandler = {[weak self] _,_  in
        self?.sendMessage()
    }

    private lazy var memePackagesView = LRSMemePackagesView(itemHandler: itemHandler, deleteHandler: deleteHandler, confirmHandler: confirmHandler).then {
        $0.backgroundColor = .color(named: "Color_244")
    }

    private var mode: Mode = .normal {
        willSet {
            toolBar.faceButton.isSelected = newValue == .meme
        }
    }
    private var memeBoardHeight: CGFloat {
        return memePackagesView.boardHeight
    }
    private let uiConfigure = BarConfigure()
    private var audioPermission = true

    var textViewHeight: CGFloat {
        let textView = toolBar.inputTextView
        let height = textView.sizeThatFits(textView.bounds.size).height
        let minHeight = configure.textView.minHeight
        let maxHeight = configure.textView.maxHeight
        return min(maxHeight, max(minHeight, height) + configure.textView.topMargin * 2)
    }

    @objc public init(frame: CGRect, configure: LRSMessageToolBarConfigure) {
        self.configure = configure
        toolBar = LRSMessageInputBar(frame: .zero, configure: configure)
        super.init(frame: frame)
        buildUI()
        addObservers()
    }

    required init?(coder: NSCoder) {
        configure = .default
        toolBar = LRSMessageInputBar(frame: .zero, configure: configure)
        super.init(coder: coder)
        buildUI()
        addObservers()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        print("LRSMessageBar -- dealloc")
    }

    @objc public func append(text: String?) {
        guard let value = text else {
            return
        }
        toolBar.inputTextView.text += value
        NotificationCenter.default.post(name: UITextView.textDidChangeNotification, object: toolBar.inputTextView)
        let offset = toolBarPosition()
        plus(offset: offset)
        memePackagesView.y(to: memePackagesView.frame.origin.y - offset)
        toolBar.scrollToEnd()
    }

    @objc @discardableResult public override func resignFirstResponder() -> Bool {
        if mode == .keyboard {
            toolBar.inputTextView.resignFirstResponder()
        } else if mode == .meme {
            animationHiddenMemePackagesView()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {[weak self] in
                self?.animationResignFirstResponder(duration: 0.3)
            }
        }
        mode = .normal
        return super.resignFirstResponder()
    }

    @objc @discardableResult public override func becomeFirstResponder() -> Bool {
        if mode == .normal {
            if toolBar.mode == .input {
                toolBar.inputTextView.becomeFirstResponder()
            } else {

            }
        } else if mode == .meme {
            onSwithMemeMode(button: toolBar.faceButton)
        }
        return true
    }

    private func buildUI() {
        toolBar.frame = CGRect(x: 0, y: 0, width: UIScreen.main.width, height: textViewHeight)
        addSubview(toolBar)

        memePackagesView.buildUI(configures: LRSMessageToolBarHelper.emojis() ?? LRSMemePackageConfigure(emojis: [], columnCount: 9))

        toolBar.recordingBtn.touchBegan = { [unowned self] button in
            if let permission = delegate?.audioPermission?() {
                audioPermission = permission
            }
            guard audioPermission == true else {
                toolBar.recordingBtn.isSelected = false
                return
            }
            delegate?.messageToolBarBeganToSpeak?(bar: self)
        }

        toolBar.recordingBtn.touchEnd = { [unowned self] _ in
            delegate?.messageToolBarEndSpeaking?(bar: self)
        }

        toolBar.recordingBtn.dragEnter = { [unowned self] _ in
            delegate?.messageToolBarDragEnterRecordScope?(bar: self)
        }

        toolBar.recordingBtn.dragOutside = { [unowned self] _ in
            delegate?.messageToolBarDragOutRecordScope?(bar: self)
        }

        toolBar.recordingBtn.dragOutsideRelease = { [unowned self] _ in
            delegate?.messageToolBarSlideTopToCancelRecording?(bar: self)
        }

        toolBar.faceButton.addTarget(self, action: #selector(onSwithMemeMode(button:)), for: .touchUpInside)
        toolBar.modeSwitchButton.addTarget(self, action: #selector(onSwithButtonClicked(button:)), for: .touchUpInside)
    }

    @objc func onSwithButtonClicked(button: UIButton) {
        switch toolBar.mode {
        case .record:
            mode = .normal
            toolBar.inputTextView.resignFirstResponder()
            toolBar.height(to: configure.textView.minHeight + configure.textView.topMargin * 2)
            animationHiddenMemePackagesView()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {[weak self] in
                self?.animationResignFirstResponder(duration: 0.3)
            }
        case .input:
            toolBarPosition()
            break
        }
    }

    @objc private func onSwithMemeMode(button: UIButton) {
        if mode == .normal || mode == .keyboard {
            mode = .meme
            toolBar.inputTextView.resignFirstResponder()
            animationBecomeFirstResponder(duration: uiConfigure.memeAnimationDuration, bottomHeight: self.memeBoardHeight)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {[weak self] in
                self?.animationShowMemePackagesView()
            }
        } else {
            toolBar.inputTextView.becomeFirstResponder()
            mode = .keyboard
            animationHiddenMemePackagesView()
        }
    }

    private func addObservers() {
        toolBar.inputTextView.delegate = self

        NotificationCenter.default.addObserver(forName: UIResponder.keyboardWillShowNotification, object: nil, queue: .main) {[weak self] noti in
            let info = noti.userInfo
            let duration = info?[UIResponder.keyboardAnimationDurationUserInfoKey] as? Double
            let to = info?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect
            self?.animationBecomeFirstResponder(duration: duration ?? 0.3, bottomHeight: to?.size.height ?? 100)
            self?.mode = .keyboard
        }

        NotificationCenter.default.addObserver(forName: UIResponder.keyboardWillHideNotification, object: nil, queue: .main) {[weak self] noti in
            if self?.mode == .keyboard {
                let info = noti.userInfo
                let duration = info?[UIResponder.keyboardAnimationDurationUserInfoKey] as? Double
                self?.animationResignFirstResponder(duration: duration ?? 0.3)
            }
        }
    }

    private func sendMessage() {
        let str = toolBar.inputTextView.text.replacingOccurrences(of: " ", with: "").replacingOccurrences(of: "\n", with: "").replacingOccurrences(of: "\r", with: "")
        delegate?.messageToolBarDidClickedReturn?(bar: self, text: str)
        toolBar.inputTextView.text = ""
        toolBar.inputTextView.setContentOffset(.zero, animated: true)
        let offset = toolBarPosition()
        plus(offset: offset)
        if mode == .meme {
            memePackagesView.y(to: memePackagesView.frame.origin.y - offset)
        }
        delegate?.sendOut(bar: self, text: str)
    }
}

extension LRSMessageBar: UITextViewDelegate {

    public func textViewDidBeginEditing(_ textView: UITextView) {
        delegate?.messageToolBarInputTextViewDidBeginEditing?(bar: self)
        UIView.animate(withDuration: 0.3) {
            var memeRect = self.memePackagesView.frame
            memeRect.origin.y = self.textViewHeight + self.memeBoardHeight
            self.memePackagesView.frame = memeRect
            self.toolBar.bottomLine.alpha = 0
        }
    }

    public func textViewShouldBeginEditing(_ textView: UITextView) -> Bool {
        return delegate?.messageToolBarShouldBeginEditting?(bar: self) ?? true
    }

    public func textViewDidEndEditing(_ textView: UITextView) {
        delegate?.messageToolBarInputTextViewDidEndEditing?(bar: self)
        UIView.animate(withDuration: 0.1) {
            self.toolBar.bottomLine.alpha = 1
        }
    }

    public func textViewDidChangeSelection(_ textView: UITextView) {
        guard let end = textView.selectedTextRange?.end, textView.selectedRange.length != 0 else {
            return
        }
        let r = textView.caretRect(for: end)
        let y = max(r.origin.y - textView.frame.size.height + r.size.height + 8, 0)
        if textView.contentOffset.y < y && r.origin.y != .infinity {
            textView.contentOffset = CGPoint(x: 0, y: y)
        }
    }

    public func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            sendMessage()
            return false
        }
        return true
    }

    public func textViewShouldEndEditing(_ textView: UITextView) -> Bool {
        textView.resignFirstResponder()
        return true
    }

    public func textViewDidChange(_ textView: UITextView) {
        let length = configure.textView.acceptLength
        if textView.text.count < length {

        } else {
            textView.text = String(textView.text.prefix(length))
        }
        let offset = toolBarPosition()
        plus(offset: offset)
        if mode == .meme {
            memePackagesView.y(to: memePackagesView.frame.origin.y - offset)
        }
    }

}


private extension UIView {
    func plus(offset: CGFloat) {
        var selfRect = self.frame
        selfRect.origin.y += offset
        selfRect.size.height -= offset
        self.frame = selfRect
    }

    func height(to: CGFloat) {
        var rect = self.frame
        rect.size.height = to
        self.frame = rect
    }

    func y(to: CGFloat) {
        var rect = self.frame
        rect.origin.y = to
        self.frame = rect
    }

}

private extension LRSMessageBar {

    @discardableResult func toolBarPosition() -> CGFloat {
        let offset = toolBar.bounds.size.height - textViewHeight
        toolBar.height(to: textViewHeight)
        return offset
    }

    private func animationHiddenMemePackagesView() {
        UIView.animate(withDuration: uiConfigure.memeAnimationDuration) {
            var memeRect = self.memePackagesView.frame
            memeRect.origin.y = self.textViewHeight + 30
            self.memePackagesView.frame = memeRect
            self.memePackagesView.alpha = 0
        }
    }

    private func animationResignFirstResponder(duration: TimeInterval) {
        let y = UIScreen.main.height - toolBar.bounds.size.height - LRSMessageToolBarHelper.safeAreaHeight
        var rect = frame
        rect.origin.y = y
        UIView.animate(withDuration: duration) {
            self.frame = rect
        }
    }

    private func animationShowMemePackagesView() {
        let y = textViewHeight + uiConfigure.memeAnimationOffset
        let rect = frame
        memePackagesView.frame = CGRect(x: 0, y: y, width: rect.size.width, height: memeBoardHeight)
        memePackagesView.alpha = 0
        if let _ = memePackagesView.superview {

        } else {
            addSubview(memePackagesView)
        }
        var memeRect = memePackagesView.frame
        memeRect.origin.y = textViewHeight + uiConfigure.toolBottomOffset

        UIView.animate(withDuration: uiConfigure.memeAnimationDuration) {
            self.memePackagesView.frame = memeRect
            self.memePackagesView.alpha = 1
        }
    }

    private func animationBecomeFirstResponder(duration: TimeInterval, bottomHeight: CGFloat) {
        let height = textViewHeight + bottomHeight + uiConfigure.toolBottomOffset
        let y = UIScreen.main.height - height
        var rect = frame
        rect.origin.y = y
        rect.size.height = height
        UIView.animate(withDuration: uiConfigure.memeAnimationDuration) {
            self.frame = rect
        }
    }
}

